_posts/2022-03-08-stream-processing-part2.md

```diff
@@ -0,0 +1,518 @@
+  val OneMinute = Duration.standardMinutes(1L)
+
+  val activities = testStreamOf[(User, Activity)]
+    .addElementsAtTime("00:00:00", ("joe", "0")).advanceProcessingTime(OneMinute)
+    .addElementsAtTime("00:01:00", ("joe", "1")).advanceProcessingTime(OneMinute)
+    .addElementsAtTime("00:02:00", ("joe", "2")).advanceProcessingTime(OneMinute)
+    .addElementsAtTime("00:03:00", ("joe", "3")).advanceProcessingTime(OneMinute)
+    .addElementsAtTime("00:04:00", ("joe", "4")).advanceProcessingTime(OneMinute)
+    .addElementsAtTime("00:05:00", ("joe", "5")).advanceProcessingTime(OneMinute)
+    .advanceWatermarkTo("00:20:00") // more than 00:05:00 + 10 minutes of gap
+    .addElementsAtTime("00:06:00", ("joe", "6")) // late event within allowed lateness
+    .addElementsAtTime("00:07:00", ("joe", "7")) // late event within allowed lateness
+    .advanceWatermarkToInfinity()
+
+  val results = activitiesInSessionWindow(
+    sc.testStream(activities),
+    TenMinutesGap,
+    accumulationMode = AccumulationMode.ACCUMULATING_FIRED_PANES,
+    allowedLateness = Duration.standardMinutes(10),
+    trigger = AfterWatermark
+      .pastEndOfWindow()
+      .withEarlyFirings(
+        AfterProcessingTime.pastFirstElementInPane().plusDelayOf(OneMinute))
+      .withLateFirings(
+        AfterPane.elementCountAtLeast(1))
+  )
+  (...)
+}
+```
+
+* In addition to the watermark, the processing time is advanced as well to fire early speculative results
+* To make a speculative results more meaningful, the `ACCUMULATING_FIRED_PANES` mode is configured
+
+Early, speculative results:
+
+```scala
+results.withTimestamp should inEarlyPane("00:00:00", "00:11:00") {
+  containSingleValueAtTime("00:10:59.999", 
+    ("joe", Iterable("0", "1")))
+}
+results.withTimestamp should inEarlyPane("00:00:00", "00:13:00") {
+  containSingleValueAtTime("00:12:59.999", 
+    ("joe", Iterable("0", "1", "2", "3")))
+}
+results.withTimestamp should inEarlyPane("00:00:00", "00:15:00") {
+  containSingleValueAtTime("00:14:59.999", 
+    ("joe", Iterable("0", "1", "2", "3", "4", "5")))
+}
+```
+
+* The first speculative result is produced one minute after "0" activity, so it contains "0" and "1" activities
+* The first pane emission resets the trigger, so the second speculative result is again produced 1 minute after "2" activity and contains activities from "0" to "3"
+* There is also a third early pane, with all already observed activities
+* The event time of the results is always the oldest observed event time plus the configured gap of ten minutes
+
+On-time result:
+
+```scala
+results.withTimestamp should inOnTimePane("00:00:00", "00:15:00") {
+  containSingleValueAtTime("00:14:59.999", 
+    ("joe", Iterable("0", "1", "2", "3", "4", "5")))
+}
+```
+
+* The on-time pane is produced when watermark has passed
+* Window and elements times are exactly like for the latest early-pane, because there has not been any new activity observed since the late pane was emitted
```

Comment:
```suggestion
* Window and elements times are exactly like for the latest early-pane, because there has not been any new activity observed since the last early-pane was emitted
```